storage/storage_reference.go

```diff
@@ -108,6 +109,23 @@ func (s storageReference) DeleteImage(ctx *types.SystemContext) error {
 		logrus.Errorf("reference %q does not resolve to an image ID", s.StringWithinTransport())
 		return ErrNoSuchImage
 	}
+	img, err := s.transport.store.GetImage(id)
+	if err != nil {
+		return errors.Wrapf(err, "error reading image %q", id)
+	}
+	if s.reference != "" {
+		nameMatch := false
+		for _, name := range img.Names {
+			if name == s.reference {
+				nameMatch = true
+				break
+			}
+		}
+		if !nameMatch {
+			logrus.Errorf("no image matching reference %q found", s.StringWithinTransport())
+			return ErrNoSuchImage
+		}
+	}
```

Comment:
Is the `resolveID`+`.GetImage`+`nameMatch` loop, which seems to be exactly the same in `newImageSource` and `DeleteImage`, likely to diverge? It might be worth sharing these 20 lines in a function.

(But it is quite fine as it is as well.)